File: fundamentals/sorting/sorting.ts
export {};

function sortArr(arr: number[]): void {
  // Legacy
  //   const asc = [...arr].sort((a, b) => a - b);
  //   const desc = [...arr].sort((a, b) => b - a);

  // New ES 2023 method
  const asc = arr.toSorted((a, b) => a - b);
  const desc = arr.toSorted((a, b) => b - a);

  console.log(`Ascending Order :`, asc);
  console.log(`Decending Order: `, desc);
}

function sortString(str: string): void {
  const asc = str.toLocaleLowerCase().split("").sort().join("");
  const desc = str.toLocaleLowerCase().split("").sort().reverse().join("");
  console.log(`Ascending Order :`, asc);
  console.log(`Decending Order: `, desc);
}

function bubbleSort(arr: number[]): number[] {
  const n = arr.length;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
      }
    }
  }
  return arr;
}

const arr = [10, 2, 1, 20];
const str = "Banana";
sortArr(arr);
sortString(str);

const array = [5, 3, 8, 4, 2];
console.log(`The bubble sort Algo: `, bubbleSort(array));

File: fundamentals/sorting/sorting.py
def sort_arr(arr):
    asc = sorted(arr)
    desc = sorted(arr, reverse = True)
    print("Ascending:", asc)
    print("Descending:", desc)
    
def sort_string(str):
    asc = "".join(sorted(str.lower()))
    desc = "".join(sorted(str.lower(), reverse=True))
    print("Ascending:", asc)
    print("Descending:", desc)
    
def bubble_sort(arr):
    n = len(arr)
    for i in range(0, n):
        for j in range(0, n-i-1):
            if(arr[j] > arr[j+1]):
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr;
                
    
arr = [10, 2, 1, 20]
str = "Banana"

sort_arr(arr)
sort_string(str)

array = [5, 3, 8, 4, 2];
print("Bubble sort Arr: ", bubble_sort(array));